core/src/main/java/org/elasticsearch/action/bulk/Retry.java

```diff
@@ -0,0 +1,226 @@
+     * @throws Exception Any exception thrown by the callable.
+     */
+    public BulkResponse withSyncBackoff(Client client, BulkRequest bulkRequest) throws Exception {
+        return SyncRetryHandler
+                .create(retryOnThrowable, backoffPolicy, client)
+                .executeBlocking(bulkRequest)
+                .actionGet();
+    }
+
+    static class AbstractRetryHandler implements ActionListener<BulkResponse> {
+        private final Client client;
+        private final ActionListener<BulkResponse> listener;
+        private final Iterator<TimeValue> backoff;
+        private final ScheduledExecutorService s = new ScheduledThreadPoolExecutor(1);
+        private final Class<? extends Throwable> retryOnThrowable;
+        private final List<BulkItemResponse> responses = new ArrayList<>();
+        private final long startTimestampNanos;
+        // needed to construct the next bulk request based on the response to the previous one
+        // volatile as we're called from a scheduled thread
+        private volatile BulkRequest currentBulkRequest;
+
+        public AbstractRetryHandler(Class<? extends Throwable> retryOnThrowable, BackoffPolicy backoffPolicy, Client client, ActionListener<BulkResponse> listener) {
+            this.retryOnThrowable = retryOnThrowable;
+            this.backoff = backoffPolicy.iterator();
+            this.client = client;
+            this.listener = listener;
+            // in contrast to System.currentTimeMillis(), nanoTime() uses a monotonic clock under the hood
+            this.startTimestampNanos = System.nanoTime();
+        }
+
+        @Override
+        public void onResponse(BulkResponse bulkItemResponses) {
+            if (!bulkItemResponses.hasFailures()) {
+                // we're done here, include all responses
+                addResponses(bulkItemResponses, (r -> true));
+                BulkResponse aggregatedResults = getCompleteResponse();
+                notifyListener(aggregatedResults);
+            } else {
+                if (canRetry(bulkItemResponses)) {
+                    addResponses(bulkItemResponses, (r -> !r.isFailed()));
+                    retry(createBulkRequestForRetry(bulkItemResponses));
+                } else {
+                    addResponses(bulkItemResponses, (r -> true));
+                    BulkResponse aggregatedResults = getCompleteResponse();
+                    notifyListener(aggregatedResults);
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            try {
+                listener.onFailure(e);
+            } finally {
+                s.shutdownNow();
+            }
+        }
+
+        private void retry(BulkRequest bulkRequestForRetry) {
+            if (backoff.hasNext()) {
+                s.schedule((() -> this.execute(bulkRequestForRetry)), backoff.next().millis(), TimeUnit.MILLISECONDS);
+            }
```

Comment:
The more we talk about it, the more I sense we're talking about different things. :)

I think all the code is quite well encapsulated in `AbstractRetryHandler#onResponse()` (I've annotated it with a few comments below):

```java
public void onResponse(BulkResponse bulkItemResponses) {
            if (!bulkItemResponses.hasFailures()) {
                // Handle success case - just return all responses gathered so far
                addResponses(bulkItemResponses, (r -> true));
                finishHim();
            } else {
                // We've got some kind of failure in at least one of the item responses.
                if (canRetry(bulkItemResponses)) {
                   // We are able to retry, so retry all failed reponses and memorize all that were ok...
                    addResponses(bulkItemResponses, (r -> !r.isFailed()));
                    retry(createBulkRequestForRetry(bulkItemResponses));
                } else {
                    // ... we cannot retry because we either the backoff policy doesn't let us or we got at least one failure which is not considered retriable. Bail out.
                    addResponses(bulkItemResponses, (r -> true));
                    finishHim();
                }
            }
        }
```

The rest is just implementation details in the individual methods that are called.